staging/src/k8s.io/apiserver/pkg/storage/etcd3/compact.go

```diff
@@ -32,44 +34,88 @@ const (
 
 var (
 	endpointsMapMu sync.Mutex
-	endpointsMap   map[string]struct{}
+	endpointsMap   map[string]*compactor
 )
 
 func init() {
-	endpointsMap = make(map[string]struct{})
+	endpointsMap = make(map[string]*compactor)
 }
 
 // StartCompactor starts a compactor in the background to compact old version of keys that's not needed.
 // By default, we save the most recent 5 minutes data and compact versions > 5minutes ago.
 // It should be enough for slow watchers and to tolerate burst.
 // TODO: We might keep a longer history (12h) in the future once storage API can take advantage of past version of keys.
-func StartCompactor(ctx context.Context, client *clientv3.Client, compactInterval time.Duration) {
+func StartCompactor(client *clientv3.Client, compactInterval time.Duration) *compactor {
 	endpointsMapMu.Lock()
 	defer endpointsMapMu.Unlock()
 
 	// In one process, we can have only one compactor for one cluster.
 	// Currently we rely on endpoints to differentiate clusters.
 	for _, ep := range client.Endpoints() {
-		if _, ok := endpointsMap[ep]; ok {
+		if c, ok := endpointsMap[ep]; ok {
 			klog.V(4).Infof("compactor already exists for endpoints %v", client.Endpoints())
-			return
+			return c
 		}
 	}
+	c := newCompactor(client, compactInterval, clock.RealClock{})
 	for _, ep := range client.Endpoints() {
-		endpointsMap[ep] = struct{}{}
+		endpointsMap[ep] = c
 	}
+	return c
+}
 
+func newCompactor(client *clientv3.Client, compactInterval time.Duration, clock clock.Clock) *compactor {
+	c := &compactor{
+		client:   client,
+		interval: compactInterval,
+		stop:     make(chan struct{}),
+		clock:    clock,
+	}
 	if compactInterval != 0 {
-		go compactor(ctx, client, compactInterval)
+		c.wg.Add(1)
+		go func() {
+			defer c.wg.Done()
+			c.runCompactLoop(c.stop)
+		}()
 	}
+	return c
+}
+
+type compactor struct {
+	client *clientv3.Client
+	wg     sync.WaitGroup
+	stop   chan struct{}
+	clock  clock.Clock
+
+	mux             sync.Mutex
+	compactRevision int64
+	interval        time.Duration
+}
+
+func (c *compactor) Stop() {
+	close(c.stop)
+	c.client.Close()
```

Comment:
Why do we close the client passed in to `StartCompactor`?  I was expecting this to be caller owned

Reply:
Reverted changes to client close.